MultiBuild/BuildLogic.cs

```diff
@@ -202,230 +237,502 @@ public static bool PlayerAction_Build_BuildMainLogic_Prefix(ref PlayerAction_Bui
+                )
+            {
+                lastRunOriginal = true;
+                return true;
+            }
+
+            __instance.waitConfirm = __instance.cursorValid;
+            __instance.multiLevelCovering = false;
+            if (__instance.handPrefabDesc != null && __instance.handPrefabDesc.multiLevel)
+            {
+                int objectProtoId = __instance.GetObjectProtoId(__instance.castObjId);
+                if (objectProtoId == __instance.handItem.ID)
                 {
-                    // no update necessary
-                    runUpdate = false;
-                    return false;
+                    __instance.multiLevelCovering = true;
                 }
-                lastPosition = __instance.groundSnappedPos;
-                forceRecalculation = false;
-
-                List<BuildPreview> previews = new List<BuildPreview>();
+            }
 
-                if (MultiBuild.IsMultiBuildRunning())
+            if (!MultiBuild.IsMultiBuildRunning() && (__instance.multiLevelCovering || !BlueprintManager.hasData))
+            {
+                if (!lastRunOriginal)
                 {
-                    if (!BlueprintManager.hasData)
-                    {
-                        BlueprintManager.data.copiedBuildings.Add(new BuildingCopy()
-                        {
-                            itemProto = __instance.handItem,
-                            recipeId = __instance.copyRecipeId,
-                            modelIndex = __instance.handPrefabDesc.modelIndex
-                        });
-                    }
-                    var building = BlueprintManager.data.copiedBuildings[0];// BlueprintManager.data.copiedBuildings.First();
+                    __instance.ClearBuildPreviews();
+                }
+                lastRunOriginal = true;
+                return true;
+            }
 
-                    int snapPath = path;
-                    Vector3[] snaps = new Vector3[1024];
+            // full hijacking of DetermineBuildPreviews
+            lastRunOriginal = false;
+            if (VFInput._switchSplitter.onDown)
+            {
+                __instance.modelOffset++;
+                forceRecalculation = true;
+            }
 
-                    var snappedPointCount = __instance.planetAux.SnapLineNonAlloc(MultiBuild.startPos, __instance.groundSnappedPos, ref snapPath, snaps);
+            if (VFInput._rotate.onDown)
+            {
+                __instance.yaw += 90f;
+                __instance.yaw = Mathf.Repeat(__instance.yaw, 360f);
+                __instance.yaw = Mathf.Round(__instance.yaw / 90f) * 90f;
+                forceRecalculation = true;
+            }
+            if (VFInput._counterRotate.onDown)
+            {
+                __instance.yaw -= 90f;
+                __instance.yaw = Mathf.Repeat(__instance.yaw, 360f);
+                __instance.yaw = Mathf.Round(__instance.yaw / 90f) * 90f;
```

Comment:
calculate this in a local variable and only set it once?

Reply:
that's original game code. i prefer to keep it like that to ensure that if they change something i can quickly spot the changes